### ISSUE TITLE
[implemented] Fractional Max Pooling

### ISSUE THREAD

**Issue by soumith** [torch/nn#371]
We've implemented fractional max pooling for both CPU and GPU. Will sync upstream in the coming few days.

**Comment by andreaskoepf** [comment]
Great news. For which of the max-pooling operations did you create a stochastic fractional version? I am especially interested in a fractional variant of SpatialAdaptiveMaxPooling since here the rounding with lots of different input sizes is very prominent.
BTW does anybody know if there is already a paper discussing fractional adaptive pooling (e.g as a variant of spatial pyramid pooling/RoI pooling)?

**Comment by soumith** [comment]
we've only implemented it for the regular SpatialMaxPooling.

**Comment by andreaskoepf** [comment]
Would it be easil

**Comment by andreaskoepf** [comment]
Would it perhaps be possible to create a variant that is parametrized on the desired output size of the pooling operation instead of x,y input-scaling factors? This would allow the use of fractional pooling for spatial pyramid pooling or RoI layers. If you have no time to work on this I would love to give it a try.

**Comment by andreaskoepf** [comment]
Original FMP seems to be defined only for a factor a with 1 < a < 2. I guess generic up/downsampling would require more work.

**Comment by soumith** [comment]
merged via: 

https://github.com/torch/nn/pull/442
https://github.com/torch/cunn/pull/155

**Comment by mattdns100689** [comment]
hi, I am getting an error with the FMP layers but not exactly sure why;

install/share/lua/5.1/nn/SpatialFractionalMaxPooling.lua:134: bad argument #3 to 'SpatialFractionalMaxPooling_updateGradInput' (gradOutput height unexpected)

does anyone have any ideas? I can get a CNN working with one conv/fmp combo but as soon as I add more I get this error.....

**Comment by soumith** [comment]
@mattdns100689 can you give a small repro script?

**Comment by mattdns100689** [comment]
Did you want the script of the making of the model. This is the model in torch's printed view;

nn.Sequential {
  [input -> (1) -> (2) -> (3) -> (4) -> (5) -> (6) -> (7) -> (8) -> (9) -> (10) -> (11) -> (12) -> (13) -> (14) -> (15) -> (16) -> (17) -> (18) -> (19) -> output]
  (1): nn.SpatialConvolutionMM(3 -> 96, 3x3, 1,1, 1,1)
  (2): nn.Tanh
  (3): nn.SpatialFractionalMaxPooling(0,0,2,2)
  (4): nn.SpatialConvolutionMM(96 -> 192, 3x3, 1,1, 1,1)
  (5): nn.Tanh
  (6): nn.SpatialFractionalMaxPooling(0,0,2,2)
  (7): nn.SpatialConvolutionMM(192 -> 288, 3x3, 1,1, 1,1)
  (8): nn.Tanh
  (9): nn.SpatialFractionalMaxPooling(0,0,2,2)
  (10): nn.SpatialConvolutionMM(288 -> 384, 3x3, 1,1, 1,1)
  (11): nn.Tanh
  (12): nn.SpatialFractionalMaxPooling(0,0,2,2)
  (13): nn.SpatialConvolutionMM(384 -> 480, 3x3, 1,1, 1,1)
  (14): nn.Tanh
  (15): nn.SpatialFractionalMaxPooling(0,0,2,2)
  (16): nn.SpatialConvolutionMM(480 -> 576, 3x3, 1,1, 1,1)
  (17): nn.View
  (18): nn.Linear(9216 -> 10)
  (19): nn.SoftMax
}

**Comment by soumith** [comment]
yes indeed, the script to make the model.

**Comment by mattdns100689** [comment]
Sorry its a bit messy. Its two scripts I use the first to make generic layer functions. The second script is the making of this particular model;

----------------------
----- First script 
require 'nn'

local layers = {}

        ----------------------------------------------------------------------------------------------------------------------
        --Define layer functions to add CNN layer or add MP layer to take above paramters
        ----------------------------------------------------------------------------------------------------------------------
function layers.add_cnn(model,cnn_layer_no)
    --First check to see if we are using the first layer;
    if cnn_layer_no > 1 then
        input_size0 = cnn_filters[cnn_layer_no-1]
    else 
        input_size0 = 3 --RGB
    end 
    return model:add(nn.SpatialConvolutionMM(input_size0,cnn_filters[cnn_layer_no],
            cnn_filter_size[cnn_layer_no],cnn_filter_size[cnn_layer_no],
        cnn_stride[cnn_layer_no],cnn_stride[cnn_layer_no],cnn_padding[cnn_layer_no],cnn_padding[cnn_layer_no]))
end


function layers.add_mp(model,mp_layer_no) 
    return model:add(nn.SpatialMaxPooling(mp_filter_size[mp_layer_no],mp_filter_size[mp_layer_no],mp_stride[mp_layer_no],
        mp_stride[mp_layer_no],mp_padding[mp_layer_no],mp_padding[mp_layer_no]))
end

function layers.add_fmp(model,layer_no) 
    return model:add(nn.SpatialFractionalMaxPooling(fmp_filter_size[layer_no],fmp_filter_size[layer_no],
           fmp_output_ratio[layer_no],fmp_output_ratio[layer_no]))
end

function layers.add_af(model,af)
    --return model:add(nn.Tanh())
    --return model:add(nn.Sigmoid())
    --return model:add(nn.PReLU())
    return model:add(af)

end

return layers

--------------------------------------------
----- Second script to make model 

layers = require "layers"

function models.fmp1()

        af = nn.Tanh()

        --final_af = nn.Softmax()

        local nFeats = 96
        cnn_filters = torch.Tensor{nFeats,nFeats*2,nFeats*3,nFeats*4,nFeats*5,nFeats*6}
        cnn_filter_size = {3,3,3,3,3,3,3}
        cnn_stride = {1,1,1,1,1,1,1}
        cnn_padding = {1,1,1,1,1,1,1}

        fmp_filter_size = {2,2,2,2,2,2}

        ratio = 2/3
        fmp_output_ratio= {ratio,ratio,ratio,ratio,ratio,ratio}

        model = nn.Sequential()

        for i = 1, 5 do
                layers.add_cnn(model,i)
                layers.add_af(model,af)
                layers.add_fmp(model,i)
                --model:add(nn.SpatialBatchNormalization(cnn_filters[i]))
        end

        layers.add_cnn(model,cnn_filters:size()[1])
        --model:add(nn.SpatialBatchNormalization(cnn_filters[{-1}]))

        local n_outputs_dense = cnn_filters[{-1}]*16
        model:add(nn.View(n_outputs_dense))
        model:add(nn.Linear(n_outputs_dense,10))
        model:add(nn.SoftMax())
        --layers.add_af(model,final_af)

        return model
end

**Comment by mattdns100689** [comment]
As a comparison this is the model that works;

```
function models.fmp0()

        --simple FMP to get working

        af = nn.Tanh()

        --final_af = nn.Softmax()

        local nFeats = 3
        cnn_filters = torch.Tensor{nFeats,nFeats*2}
        cnn_filter_size = {3,3}
        cnn_stride = {1,1}
        cnn_padding = {1,1}

        fmp_filter_size = {2,2,2,2,2,2}

        ratio = 2/3

        fmp_output_ratio= {ratio,ratio,ratio,ratio,ratio,ratio}

        model = nn.Sequential()

        for i = 1, 1 do
                layers.add_cnn(model,i)
                layers.add_af(model,af)
                layers.add_fmp(model,i)
                model:add(nn.SpatialBatchNormalization(cnn_filters[i]))
        end

        layers.add_cnn(model,cnn_filters:size()[1])
        model:add(nn.SpatialBatchNormalization(cnn_filters[{-1}]))

        local n_outputs_dense = cnn_filters[{-1}]*21*21
        model:add(nn.View(n_outputs_dense))
        model:add(nn.Linear(n_outputs_dense,10))
        model:add(nn.SoftMax())
        --layers.add_af(model,final_af)

        return model
end
```

**Comment by mattdns100689** [comment]
So I managed to get it rid of the error by explicitly stating the output sizes for width/height. e.g;

```
th> model:cuda()
nn.Sequential {
  [input -> (1) -> (2) -> (3) -> (4) -> (5) -> (6) -> (7) -> (8) -> (9) -> (10) -> (11) -> (12) -> (13) -> (14) -> (15) -> (16) -> (17) -> (18) -> output]
  (1): nn.SpatialConvolutionMM(3 -> 100, 3x3, 1,1, 1,1)
  (2): nn.Tanh
  (3): nn.SpatialFractionalMaxPooling(21,21,2,2)
  (4): nn.SpatialConvolutionMM(100 -> 200, 3x3, 1,1, 1,1)
  (5): nn.Tanh
  (6): nn.SpatialFractionalMaxPooling(14,14,2,2)
  (7): nn.SpatialConvolutionMM(200 -> 300, 3x3, 1,1, 1,1)
  (8): nn.Tanh
  (9): nn.SpatialFractionalMaxPooling(9,9,2,2)
  (10): nn.SpatialConvolutionMM(300 -> 400, 3x3, 1,1, 1,1)
  (11): nn.Tanh
  (12): nn.SpatialFractionalMaxPooling(6,6,2,2)
  (13): nn.SpatialConvolutionMM(400 -> 500, 3x3, 1,1, 1,1)
  (14): nn.Tanh
  (15): nn.SpatialFractionalMaxPooling(4,4,2,2)
  (16): nn.View
  (17): nn.Linear(8000 -> 10)
  (18): nn.SoftMax
}
```
However I have run into another error now; 

```
cuda runtime error (77) : an illegal memory access was encountered at /home/msmith/torch/extra/cutorch/lib/THC/THCTensorMath.cu:318
```

**Comment by soumith** [comment]
cc: @colesbury you know a little bit about the fractional max pooling implementation. Do you know what's happening here?

**Comment by mattdns100689** [comment]
So I took everything off the GPU onto the CPU and get this;
```
...install/share/lua/5.1/nn/SpatialFractionalMaxPooling.lua:134: Assertion `index >= 0 && index < inputW * inputH' failed. 
```

Not to sure if this helps.

**Comment by mattdns100689** [comment]
@soumith This was a silly error in the end, I just added af = nn.Tanh() to each layer when I needed to call a different one each time.

**Comment by soumith** [comment]
@mattdns100689 good to know. I missed that when looking at the reference script.
